### ISSUE TITLE
Parse RPC cannot handle file contents that are invalid UTF-8.

### ISSUE THREAD

**Issue by creachadair** [bblfsh/bblfshd#238]
## Summary

This issue was reported by Fábio Costa Silva in the [source{d} community Slack](https://sourced-community.slack.com). What follows is my distillation of it and a relatively-minimal repro.

The [`ParseRequest`](https://github.com/bblfsh/sdk/blob/42e2023d304c66a242444d467c48f6913748e411/protocol/driver.proto#L14) protobuf message specifies a `content` field of type `string`. Per the [protobuf language guide](https://developers.google.com/protocol-buffers/docs/proto3#scalar),

> A string must always contain UTF-8 encoded or 7-bit ASCII text.

If a file is encoded in some other way, e.g., ISO-8859-1 (Latin 1), the RPC layer will generate an error marshaling the request message and fail.

## Repro

```shell
# One-time setup
mkdir repro
cd repro
git init
git commit -m "Initial commit." --allow-empty
python -c "print u'package bad;\npublic class Bad {\n  static final String bad = \"\\xe9\";\n}\n'.encode('latin-1')" > Bad.java
git add Bad.java
git commit -m "Add a non-UTF-8 encoded Java file." .
srcd init

# Reproduction
srcd sql 'select uast(blob_content, language(file_path, blob_content)) from files;'
```
## Expected

A UAST for this syntactically-valid Java file.

## Observed

```
rpc error: code = Internal desc = grpc: error while marshaling: proto: invalid UTF-8 string
```

The correct fix is to convert this field to type `bytes`. That is technically a breaking change, but in this case it is a fairly innocuous one: Any text that would have been accepted by a `string` field will also be accepted by a `bytes` field, but not vice versa. So it is wire-safe, but it changes the API surface of the generated code.

**Comment by creachadair** [comment]
Note that this error does not occur if you run `srcd parse uast Bad.java`. I assume this is because that code path does not make an RPC, but uses the library directly (but that is a guess, I haven't looked).

**Comment by juanjux** [comment]
Currently we only support UTF-8 or ASCII (since it's a subset). The problem is not only for protobuf, but also for the SDK and bblfshd which expect UTF-8, so just converting to bytes wouldn't work. Note that ISO-8859-1 is not a subset of UTF-8 and thus is incompatible so its up to the user to convert any source code into UTF-8 before passing it to bblfsh. What we could do is to do the conversion ourselves if the user specifies the encoding (because guessing it from the source file doesn't always work).

**Comment by creachadair** [comment]
> [I]t's up to the user to convert any source code into UTF-8 before passing it to bblfsh. 

Is that documented anywhere? At the very least we should report a better error than a marshaling failure when this happens.

Guessing the encoding is clearly the wrong solution, but I had in mind to allow the user to specify the encoding as a separate field.

**Comment by dennwc** [comment]
Agree with @juanjux, all the pipeline expects UTF8. We should convert the string before sending it to the drivers.

We can extend the protocol for `bblfshd` and allow to specify an encoding. `bblfshd` will take care of the conversion and will send UTF8 to the driver. We do exactly the same for v2 -> v1 compatibility layer.

**Comment by creachadair** [comment]
If Babelfish requires UTF-8, I think we need to document that very clearly and give a better diagnostic message than a random failure from inside the stack. I agree that the priority of supporting other encodings is low, but requests should report an orderly error instead of failing IMO.

**Comment by dennwc** [comment]
Sounds good, let's check the content for UTF8.

**Comment by fcostasilva** [comment]
Hi All, 
After trying several projects, I believe the root cause is not an invalid encoding. At least Linux and Mac show "us-ascii" whenever I check a file. If I try to navigate through the AST I get an even worse error. For example:

statement:
`UAST(files.blob_content, LANGUAGE(files.file_path, files.blob_content), '//*[@roleCallee]/Identifier[@Name="trim"]')`

result:

`rpc error: code = Unknown desc = SQL query failed: Error 1105: unknown error: invalid type: BLOB`

Based on this behaviour, I have two questions:
- Is there a way to show the error with more details? 
- Do you have a git project that is known for working fine with scrd? If so, I could use it to verify if it's a local problem or an actual defect.

**Comment by creachadair** [comment]
> After trying several projects, I believe the root cause is not an invalid encoding. At least Linux and Mac show "us-ascii" whenever I check a file.

Most of the files in that repository are indeed US-ASCII or UTF-8, but there are a dozen or so that are other things (mostly fragments, and in one case ISO-8859-1). I don't actually know which file triggered the original error (I couldn't see the log output from inside the container), but I am confident they're not all plain ASCII.

> `rpc error: code = Unknown desc = SQL query failed: Error 1105: unknown error: invalid type: BLOB`

That appears like it may be a different error than the one this issue describes.

> Based on this behaviour, I have two questions:
> 
> * Is there a way to show the error with more details?
> * Do you have a git project that is known for working fine with scrd? If so, I could use it to verify if it's a local problem or an actual defect.

@ajnavarro have you any suggestions here?

In general my expectation is that things should be working, but there are clearly some issues with this repo.

**Comment by fcostasilva** [comment]
As a matter of fact, I have just executed the same query in the web client and the command-line, and the web client works fine.

command-line:
![command-line](https://user-images.githubusercontent.com/7120328/52090111-b6a13200-25b0-11e9-92dd-73eb0fef1583.png)

web-client:
![web-client](https://user-images.githubusercontent.com/7120328/52090125-c1f45d80-25b0-11e9-9f86-558c15d5d61c.png)

.

**Comment by se7entyse7en** [comment]
I think that this is the same issue that was open on [engine](https://github.com/src-d/engine/issues/196).

**Comment by ajnavarro** [comment]
@creachadair @fcostasilva 

To know the cause of `invalid type: BLOB` we need the complete query. 

On that query, some blob column is used in some place that is not possible. That part: `UAST(files.blob_content, LANGUAGE(files.file_path, files.blob_content), '//*[@roleCallee]/Identifier[@Name="trim"]')` appears to be OK.

**Comment by creachadair** [comment]
> @creachadair @fcostasilva
> 
> To know the cause of `invalid type: BLOB` we need the complete query.
> 
> On that query, some blob column is used in some place that is not possible. That part: `UAST(files.blob_content, LANGUAGE(files.file_path, files.blob_content), '//*[@roleCallee]/Identifier[@Name="trim"]')` appears to be OK.

I believe the OP's query is more or less one of the examples from the documentation, so I am hoping it should be OK.

**Comment by fcostasilva** [comment]
@creachadair Yes, this is the example "Find all files where 'trim' method is called" from the documentation. This one also fails on the web client, as you can see in the attachment. 

![web](https://user-images.githubusercontent.com/7120328/52379627-44769480-2a6b-11e9-9c12-f18e6843e6a7.png)

![trim](https://user-images.githubusercontent.com/7120328/52378737-c7e2b680-2a68-11e9-897d-e5d82fb2d5bc.png)

**Comment by ajnavarro** [comment]
the `invalid type: BLOB` is because the query you are using is not compatible anymore with the way we are serializing the bblfsh output. Before it was a JSON array, now it is a blob.

`ARRAY_LENGTH(JSON)` expects a JSON, not a blob. The output from `UAST` is a blob with a special serialization to make it more performant.

To make that query works with an actual gitbase version, you need to add `uast_extract` ([more info here](https://docs.sourced.tech/gitbase/using-gitbase/functions#how-to-use-uast_extract)):

```sql
explain format=tree select * from (
	select
		files.repository_id,
		files.file_path,
		UAST(files.blob_content, LANGUAGE(files.file_path, files.blob_content), '//*[@roleCallee]/Identifier[@Name="trim"]') as functionCall
	from files
	natural join commit_files
	natural join commits
	natural join refs
	where refs.ref_name = "HEAD"
) t where array_length(uast_extract(functionCall,'@type')) > 0
```

**Comment by carlosms** [comment]
https://github.com/src-d/engine/issues/196 has been included in the [engine v0.10.0](https://github.com/src-d/engine/releases/tag/v0.10.0) release. This fixes the `invalid UTF-8 string` error between the `srcd sql` client and the `srcd-cli-daemon` container. I don't know if this encoding issue is also present in some other internal bblfsh communication, though...

**Comment by creachadair** [comment]
> [src-d/engine#196](https://github.com/src-d/engine/issues/196) has been included in the [engine v0.10.0](https://github.com/src-d/engine/releases/tag/v0.10.0) release. This fixes the `invalid UTF-8 string` error between the `srcd sql` client and the `srcd-cli-daemon` container. I don't know if this encoding issue is also present in some other internal bblfsh communication, though...

The driver protocol does contain a `string` field which could in principle be affected, however I was not able to reproduce it.
